server/core/src/main/java/org/infinispan/server/core/dataconversion/ProtostreamTranscoder.java

```diff
@@ -0,0 +1,146 @@
+import static org.infinispan.commons.dataconversion.MediaType.APPLICATION_OBJECT;
+import static org.infinispan.commons.dataconversion.MediaType.APPLICATION_OCTET_STREAM;
+import static org.infinispan.commons.dataconversion.MediaType.APPLICATION_PROTOSTREAM;
+import static org.infinispan.commons.dataconversion.MediaType.APPLICATION_UNKNOWN;
+import static org.infinispan.commons.dataconversion.MediaType.TEXT_PLAIN;
+import static org.infinispan.commons.dataconversion.StandardConversions.convertTextToObject;
+import static org.infinispan.commons.dataconversion.StandardConversions.decodeOctetStream;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.io.StringReader;
+import java.util.Optional;
+
+import org.infinispan.commons.dataconversion.MediaType;
+import org.infinispan.commons.dataconversion.OneToManyTranscoder;
+import org.infinispan.commons.dataconversion.StandardConversions;
+import org.infinispan.commons.marshall.ProtoStreamMarshaller;
+import org.infinispan.commons.marshall.WrappedByteArray;
+import org.infinispan.commons.util.Util;
+import org.infinispan.protostream.ProtobufUtil;
+import org.infinispan.protostream.SerializationContext;
+import org.infinispan.util.logging.Log;
+import org.infinispan.util.logging.LogFactory;
+
+/**
+ * Transcode between application/x-protostream and commons formats
+ *
+ * @since 10.0
+ */
+public class ProtostreamTranscoder extends OneToManyTranscoder {
+
+   protected final static Log logger = LogFactory.getLog(JBossMarshallingTranscoder.class, Log.class);
+   private volatile SerializationContext ctx;
+   private final ProtoStreamMarshaller marshaller;
+   private final ClassLoader classLoader;
+
+   public ProtostreamTranscoder(SerializationContext ctx, ClassLoader classLoader) {
+      super(APPLICATION_PROTOSTREAM, APPLICATION_OCTET_STREAM, TEXT_PLAIN, APPLICATION_OBJECT, APPLICATION_JSON, APPLICATION_UNKNOWN);
+      this.ctx = ctx;
+      this.marshaller = new ProtoStreamMarshaller(ctx);
+      this.classLoader = classLoader;
+   }
+
+   public void setCtx(SerializationContext ctx) {
+      this.ctx = ctx;
+   }
+
+   @Override
+   public Object transcode(Object content, MediaType contentType, MediaType destinationType) {
+      try {
+         if (destinationType.match(MediaType.APPLICATION_PROTOSTREAM)) {
+            Object decoded = content;
+            if (contentType.match(APPLICATION_OBJECT)) {
+               decoded = StandardConversions.decodeObjectContent(content, contentType);
+               Optional<String> wrappedParam = destinationType.getParameter("wrapped");
+               if (!wrappedParam.isPresent() || !wrappedParam.get().equals("false"))
+                  return ProtobufUtil.toWrappedByteArray(ctx, decoded);
+               return ProtobufUtil.toByteArray(ctx, decoded);
+            }
+            if (contentType.match(APPLICATION_OCTET_STREAM)) {
```

Comment:
wouldn't it be better to use `else if`? or the `contentType` can match multiple MediaType?

Reply:
I think when the content type is MATCH_ALL this behaviour is desirable. I used the existing JBossMarshallingTranscoder as a base for this one, which handles the content/destination types in this manner.

Reply:
I'm confused now... how can the `content` be `MATCH_ALL`? 
if that is the case, you can just do `return content` since `MATCH_ALL` will match whatever is in `destinationType`. 

ps. can't you do `if (componentType.match(destinationType) return content`? if the component matches the destination, there is no transcoding needed. 
ps2. I'll leave the review for the experts :)

Reply:
The idea is to convert a certain ```content``` from ```contentType``` to ```destinationType```, and if both media types are compatible, the transcoder is not called

Reply:
> I'm confused now... how can the `content` be `MATCH_ALL`?

Sorry, I meant the specified MediaType `contentType`, not the content itself.


> ps. can't you do `if (componentType.match(destinationType) return content`? if the component matches the destination, there is no transcoding needed.

I don't think so, because of MATCH_ALL again.